feat: introduce the turborepo to the parser.js

ayushnau opened the pull request:
Changes Added.

    Turborepo Added.
    Converted the parser repo to app, inside the turborepo.
    Added the required scripts of the turborepo to the root package.json file.

Related issue(s)

Related to: https://github.com/asyncapi/parser-js/issues/963

ayushnau commented:
Deliverables:
1. Integrate turborepo in parser.js
2. Integrate multiparser in the parser.js.

ayushnau commented:
@smoya please review the pr.

smoya commented:
/dnm

smoya commented:
Blocking the merge until we review it

smoya commented on README.md:
```diff
@@ -1,533 +1,81 @@
-[![AsyncAPI JavaScript Parser](./assets/logo.png)](https://www.asyncapi.com)
+# Turborepo starter
```
We need a proper README. This is just a placeholder.

smoya commented on apps/parser/CONTRIBUTING.md:
This file should go in the root of the repository

AyushNautiyalDeveloper replied:
added in root.

smoya commented on apps/parser/LICENSE:
This file should go in the root of the repository

smoya commented on apps/parser/.gitignore:
```diff
@@ -0,0 +1,49 @@
+.turbo
+
+# Vercel
+.vercel
```
Why adding so much things to the .gitignore? What are those additions based on?

For example, the `.next` dir?

AyushNautiyalDeveloper replied:
removed unwanted keys.

smoya commented on turbo.json:
```diff
@@ -0,0 +1,15 @@
+  "globalDependencies": ["**/.env.*local"],
+  "pipeline": {
+    "build": {
+      "outputs": [".next/**", "!.next/cache/**"]
```
Why next?

AyushNautiyalDeveloper replied:
removed.

smoya replied:
Did you test the build?

AyushNautiyalDeveloper commented on apps/parser/CODEOWNERS:
added in root.

AyushNautiyalDeveloper commented on apps/parser/CODE_OF_CONDUCT.md:
added in root.

smoya commented on README.md:
In one side, please revert this file back to the version in `master` branch. You are not making additions but just styling changes, which are not related to the work to do in this PR.

On the other side, you should have one README in the root, mentioning this repository is a monorepo that contains several apps (only parser-js atm), similar to https://github.com/asyncapi/studio. And as nice to have, I would say a link to the README's of each app.

smoya commented on apps/parser/LICENSE:
Move this file to the root of the repo.

smoya commented:
Please change the prefix of the PR to `refactor:` or `chore:`. Otherwise, we would be releasing a new version of the parser-js as if it was a feature when it's not.

Additionally, edit the `Resolves #963` part in the issue description to something like `Related to #963` since this PR is not solving it fully. Otherwise, the issue will be closed automatically once this PR gets merged.

smoya commented:
What do you think are the next steps? Shall we manually test this? Like trying to run a dry-run release.

smoya commented:
@aayushmau5 please rename the PR accordingly to https://github.com/asyncapi/parser-js/pull/992#issuecomment-2085224668

smoya commented:
@ayushnau How are we gonna build the parser from now on? Should `npm run build` just work? At this moment, it gives you the following error (it's true, no `build` task).

```bash
❯ npm run build

> build
> turbo run build

  × missing tasks in project

Error:   × could not find task `build` in project
```

Shall we additionally change any path (if it changes) or command in the GH actions workflows? Have you tried to run semantic-release in dry-run mode?

ayushnau commented:
> @ayushnau How are we gonna build the parser from now on? Should `npm run build` just work? At this moment, it gives you the following error (it's true, no `build` task).
> 
> ```shell
> ❯ npm run build
> 
> > build
> > turbo run build
> 
>   × missing tasks in project
> 
> Error:   × could not find task `build` in project
> ```
> 
> Shall we additionally change any path (if it changes) or command in the GH actions workflows? Have you tried to run semantic-release in dry-run mode?

till other repo is not merged "npm run build" is fine after that will add the command for parser only build "npm run parser:build" like this.

smoya commented:
For the record, as I shared with you @ayushnau via Slack DM, the `npm run build` command is failing on my computer. I'm using node v18. There is something I'm doing wrong? 

![image](https://github.com/asyncapi/parser-js/assets/1083296/6c40f8ff-4591-4d2d-9a88-bab28655dcd3)

smoya commented:
@ayushnau Did you have a chance to test it? 🙏

ayushnau commented:
I looked into it, 

> @ayushnau Did you have a chance to test it? 🙏

@smoya which node version are you using.  Cant find the exact reason but most probably it is happening because of node version. Mine is version 20.

smoya commented:
> I looked into it,
> 
> > @ayushnau Did you have a chance to test it? 🙏
> 
> @smoya which node version are you using. Cant find the exact reason but most probably it is happening because of node version. Mine is version 20.

No, it wasn't the version of Node or NPM, but rather I had to manually remove both `node_modules` dir from root and `apps/parser` dirs.

smoya commented:
@ayushnau I believe next step should be to emulate a release by running `npx semantic-release --dry-run`. You will require to set tokens, I recommend you create your owns. In case of pushing to npm, you could change the package scope so you can push to your own npm account.

ayushnau commented:
<img width="1007" alt="image" src="https://github.com/asyncapi/parser-js/assets/78146753/7ae49fa7-9af3-4fa0-b29a-9097914e5f4a">
<img width="1055" alt="image" src="https://github.com/asyncapi/parser-js/assets/78146753/e6c97a49-1a48-4631-9235-de9b9fc4db04">

@smoya The release is working fine in dry run.

aeworxet commented:
@asyncapi/bounty_team

aeworxet commented:
/bounty

smoya commented:
> <img width="1007" alt="image" src="https://github.com/asyncapi/parser-js/assets/78146753/7ae49fa7-9af3-4fa0-b29a-9097914e5f4a">
> <img width="1055" alt="image" src="https://github.com/asyncapi/parser-js/assets/78146753/e6c97a49-1a48-4631-9235-de9b9fc4db04">
> 
> @smoya The release is working fine in dry run. 

As you are aware @ayushnau, im off this week, so I cant merge this PR until im back.
This PR contains critical changes so it requires a controlled and watched merge and release.

smoya commented:
@jonaslagoni @magicmatatjahu would you have a chance to do a review on this one before next week?🙏

jonaslagoni reviewed (changes requested):
We have a huge regression of the documentation, root readme files etc, that should not happen with the rewrite unless it's to a temporary branch 🙂

ayushnau commented:
@smoya @jonaslagoni I have updated the readmd file. 
Also other than that I have added the multiparser in the turborepo. working on pointing the multiparser to local parser repo.

smoya commented on README.md:
```diff
@@ -1,3 +1,25 @@
+# Parser-Js
+
+### Installing
```
Please, merge all of this new docs (or remove if repeated) into the previous README contents.

As you can see, and as @jonaslagoni mentioned, we don't want to have a worse documentation than what we have. It's about the format, the header of the repository, the TOC, etc. 
![Google Chrome_jjqAjmOb](https://github.com/asyncapi/parser-js/assets/1083296/be3a4d8a-9921-4f45-855a-afe208147b87)

smoya commented:
> Also other than that I have added the multiparser in the turborepo. (ayushnau:feat-multiparser branch) working on pointing the multiparser to local parser repo.

If you want us to review, please open a PR.

smoya reviewed (changes requested):
https://github.com/asyncapi/parser-js/pull/992#discussion_r1615754713

smoya commented:
> I encourage you to do a proper release without the dry-run. In that case, before pushing to npm, you could change the package scope so you can push to your own npm account.

@ayushnau have you ever tested that?

AyushNautiyalDeveloper commented:
> > I encourage you to do a proper release without the dry-run. In that case, before pushing to npm, you could change the package scope so you can push to your own npm account.
> 
> @ayushnau have you ever tested that?

No I have done the test for only dry run.

smoya commented:
> No I have done the test for only dry run.

I feel it is totally required to do this before merging.

ayushnau commented:
@smoya updated the readme.md file. 
also here is the pr for multiparser for review https://github.com/ayushnau/parser-js/pull/1 check.

smoya commented on README.md:
```diff
@@ -15,34 +21,33 @@ Updated bundle for the browser is always attached to the GitHub Release.
 - [Installation](#installation)
 - [Usage](#usage)
 - [Examples](#examples)
-  * [Example with parsing](#example-with-parsing)
```
The TOC content is autogenerated, so please revert them back to previous values.

smoya commented on README.md:
```diff
@@ -6,6 +6,12 @@ Updated bundle for the browser is always attached to the GitHub Release.
+
+This is a Monorepo managed using [Turborepo](https://turbo.build/) and contains the following package:
+
+1. **Parser-js**: Use this package to validate and parse AsyncAPI documents —either YAML or JSON— in your Node.js or browser application.
```
```suggestion
1. [Parser-js](apps/parser): Use this package to validate and parse AsyncAPI documents —either YAML or JSON— in your Node.js or browser application.
```

smoya commented on README.md:
```diff
@@ -6,6 +6,12 @@ Updated bundle for the browser is always attached to the GitHub Release.
 ![npm](https://img.shields.io/npm/v/@asyncapi/parser?style=for-the-badge) ![npm](https://img.shields.io/npm/dt/@asyncapi/parser?style=for-the-badge)
 
+
+This is a Monorepo managed using [Turborepo](https://turbo.build/) and contains the following package:
```
I believe this should be moved up and somehow replace content from line 3 and beyond (but keeping all the details about Spectral, etc).

ayushnau commented:
<img width="968" alt="image" src="https://github.com/asyncapi/parser-js/assets/78146753/1ccda70c-5ca3-4a89-ae7f-a9ba4404d72d">
<img width="960" alt="image" src="https://github.com/asyncapi/parser-js/assets/78146753/f68cde4e-2ed3-4362-a2c8-200ac4337700">

@smoya it is working fine with 
`npx semantic-release` it is working fine but only if I am making a small change. 

<img width="654" alt="image" src="https://github.com/asyncapi/parser-js/assets/78146753/dc530cd2-278a-4088-b15c-49267278ed8b">
here there in node_modules dependency of `conventional-changelog-conventionalcommits` that is basically checking for the dates. but the dates it is recieving for all the commits is `{}` which is not getting return. hence causing error. 

So now My question making changes in the node_modules is possible here?

smoya commented:
> This dependency `conventional-changelog-conventionalcommits` is installed by me to run the for running semantic release. will be need this dependency to be present there or not.

@ayushnau I'm sorry but after reading several times I didn't really get if this is a question regarding the need of such package (we need it for conventional commits) or that you are saying the culprit is such a package. What is exactly the "small change" you refer to?

AyushNautiyalDeveloper commented:
> > This dependency `conventional-changelog-conventionalcommits` is installed by me to run the for running semantic release. will be need this dependency to be present there or not.
> 
> @ayushnau I'm sorry but after reading several times I didn't really get if this is a question regarding the need of such package (we need it for conventional commits) or that you are saying the culprit is such a package. What is exactly the "small change" you refer to?

1. for testing `npx semantic-realease` I was needed to install `conventional-changelog-conventionalcommits` package. Now is it alright or not. As this package is not the part of the parser-js. I used it just for testing purposes.
2. if yes, while using this I was required to make some changes in the node_modules of this package. Now can you give me the comment on this. Overall if this is correct or not.

ayushnau commented:
> > This dependency `conventional-changelog-conventionalcommits` is installed by me to run the for running semantic release. will be need this dependency to be present there or not.
> 
> @ayushnau I'm sorry but after reading several times I didn't really get if this is a question regarding the need of such package (we need it for conventional commits) or that you are saying the culprit is such a package. What is exactly the "small change" you refer to?

1. for testing `npx semantic-realease` I was needed to install `conventional-changelog-conventionalcommits` package. Now is it alright or not. As this package is not the part of the parser-js. I used it just for testing purposes.
2. if yes, while using this I was required to make some changes in the node_modules of this package. Now can you give me the comment on this. Overall if this is correct or not.

smoya commented:
> 1. for testing `npx semantic-realease` I was needed to install `conventional-changelog-conventionalcommits` package. Now is it alright or not. As this package is not the part of the parser-js. I used it just for testing purposes.

That package is installed by the GH action that runs the release. See https://github.com/asyncapi/parser-js/blob/05b782b1ae4019a5ecf41ab0a2984a660b40114c/.github/workflows/if-nodejs-release.yml#L110

So you don't need to install it.

> 2. if yes, while using this I was required to make some changes in the node_modules of this package. Now can you give me the comment on this. Overall if this is correct or not.

I don't expect any change is needed to be done. What are the changes you needed to do?

ayushnau commented:
add Object.keys condition 
<img width="442" alt="image" src="https://github.com/asyncapi/parser-js/assets/78146753/cf623d85-6dd5-43f8-b88f-3a040b920738">

smoya commented:
> added the  Object.keys condition in the node_modules of the package. Otherwise I am getting error. @smoya 
> 
> <img width="442" alt="image" src="https://github.com/asyncapi/parser-js/assets/78146753/cf623d85-6dd5-43f8-b88f-3a040b920738">
> 

But you mean in the `conventional-changelog-conventionalcommits` package?
We do not.own such a package. Are you sure we are doing things right? Maybe some misconfiguration?

ayushnau commented:
@smoya It is working fine now. I changed the version and now it is working. 
<img width="824" alt="image" src="https://github.com/asyncapi/parser-js/assets/78146753/bcd68be6-c84c-4d7f-8369-a1e96270043c">

smoya reviewed (dismissed):
LGTM! 🚀🌔

smoya commented:
@ayushnau There is an error, seems a not updated .lock file or similar https://github.com/asyncapi/parser-js/actions/runs/9291475481/job/25618079508?pr=992

ayushnau commented:
@smoya updated can you check now.

smoya commented:
There is [an issue with our tests](https://github.com/asyncapi/parser-js/actions/runs/9348904713?pr=999) in CI when running under macos. It is not the first time we find this. 
We need to investigate what the issue is, but not linked to this PR.

AyushNautiyalDeveloper commented:
ok sure @smoya

ayushnau commented:
@smoya please check i have updated the pr.

smoya commented:
@jonaslagoni can you give your 👍 now?

jonaslagoni reviewed (approved):
@smoya if you are happy with it, go for it 👍

smoya commented:
> @smoya if you are happy with it, go for it 👍 Dont see anything from my side

I only have one concern. I want to merge this ASAP and release it (so we can be sure everything works as expected, including CI pipeline etc). The point is that this is a `chore` and atm I don't see any opened PR with a mergeable status we could merge right after this one in order to do a meaningful release. 
Forcing a release for this without providing any new feature or fix is not very desirable... 

Any idea or suggestion @jonaslagoni ?

jonaslagoni commented:
I think it's fine regardless, do what is easiest don't think too much about it 🙂

smoya commented:
I'm gonna merge this as a new minor version (`feat`). I prefer to fail fast than cry late.

smoya commented:
/rtm

smoya commented:
> I'm gonna merge this as a new minor version (`feat`). I prefer to fail fast than cry late.

Well, failed fast 😆 

> [7:50:15 AM] [semantic-release] [@semantic-release/npm] › ℹ  Skip publishing to npm registry as package.json's private property is true

See https://github.com/asyncapi/parser-js/actions/runs/9413515248/job/25930521532#step:9:65

@ayushnau I believe you tested this as per your previous comments (to publish to your own NPM account). Anyway, we need to fix this asap as `3.1.0` is published only in GH but not in NPM :\

smoya commented:
The issue comes with semantic-release and it's lack of support for monorepo. We should move into [Changesets](https://github.com/changesets/changesets) as Studio did: https://github.com/asyncapi/studio/blob/master/doc/adr/0002-use-changesets-for-publishing-and-versioning.md

I'm reverting this PR. @ayushnau Feel free to open a new one with a version that really releases each package individually.
